qiskit/extensions/quantum_initializer/initializer.py

```diff
@@ -60,318 +55,27 @@ def __init__(self, params, num_qubits=None):
-
-            num_qubits = int(num_qubits)
-
-        super().__init__("initialize", num_qubits, 0, params)
-
-    def _define(self):
-        if self._from_label:
-            self.definition = self._define_from_label()
-        elif self._from_int:
-            self.definition = self._define_from_int()
-        else:
-            self.definition = self._define_synthesis()
-
-    def _define_from_label(self):
-        q = QuantumRegister(self.num_qubits, "q")
-        initialize_circuit = QuantumCircuit(q, name="init_def")
-
-        for qubit, param in enumerate(reversed(self.params)):
-            initialize_circuit.append(Reset(), [q[qubit]])
-
-            if param == "1":
-                initialize_circuit.append(XGate(), [q[qubit]])
-            elif param == "+":
-                initialize_circuit.append(HGate(), [q[qubit]])
-            elif param == "-":
-                initialize_circuit.append(XGate(), [q[qubit]])
-                initialize_circuit.append(HGate(), [q[qubit]])
-            elif param == "r":  # |+i>
-                initialize_circuit.append(HGate(), [q[qubit]])
-                initialize_circuit.append(SGate(), [q[qubit]])
-            elif param == "l":  # |-i>
-                initialize_circuit.append(HGate(), [q[qubit]])
-                initialize_circuit.append(SdgGate(), [q[qubit]])
-
-        return initialize_circuit
-
-    def _define_from_int(self):
-        q = QuantumRegister(self.num_qubits, "q")
-        initialize_circuit = QuantumCircuit(q, name="init_def")
-
-        # Convert to int since QuantumCircuit converted to complex
-        # and make a bit string and reverse it
-        intstr = f"{int(np.real(self.params[0])):0{self.num_qubits}b}"[::-1]
-
-        # Raise if number of bits is greater than num_qubits
-        if len(intstr) > self.num_qubits:
-            raise QiskitError(
-                "Initialize integer has %s bits, but this exceeds the"
-                " number of qubits in the circuit, %s." % (len(intstr), self.num_qubits)
-            )
-
-        for qubit, bit in enumerate(intstr):
-            initialize_circuit.append(Reset(), [q[qubit]])
-            if bit == "1":
-                initialize_circuit.append(XGate(), [q[qubit]])
-
-        return initialize_circuit
-
-    def _define_synthesis(self):
-        """Calculate a subcircuit that implements this initialization
-
-        Implements a recursive initialization algorithm, including optimizations,
-        from "Synthesis of Quantum Logic Circuits" Shende, Bullock, Markov
-        https://arxiv.org/abs/quant-ph/0406176v5
-
-        Additionally implements some extra optimizations: remove zero rotations and
-        double cnots.
-        """
-        # call to generate the circuit that takes the desired vector to zero
-        disentangling_circuit = self.gates_to_uncompute()
+        self._stateprep = StatePreparation(params, num_qubits)
```

Comment:
can you give me a more specific example? I didn't explicitly check this but the params handling should just be the same if I moved everything over correctly

Reply:
I didn't do any testing, but I was thinking that I could potentially do something pathological like
```python
from qiskit.extensions import Initialize
import numpy as np

init = Initialize(np.array([1, 1]) * np.sqrt(0.5))
init.params = np.array([1, -1]) * np.sqrt(0.5)
```
and bad things would probably happen - in the new form, the `params` won't propagate through to `StatePreparation`.  That said, it's highly likely that `Initialize` is already broken with respect to mutation of `params` in some form or another, and if so, it's not really your responsibility to fix it.

Reply:
Running this example didn't throw any errors for me so either nothing bad is happening or bad things are happening silently 🤷‍♀️

Reply:
It wouldn't throw an error just on the code I wrote, but I was thinking about things that depend on `params`.  Things like how `init.definition` behaves - both before and after this change, will it do the right thing if the parameters are mutated between object creation and first definition?  My guess is that in the original code it will, and the new code it won't.